src/targets/maven.ts

```diff
@@ -102,48 +93,115 @@ export class MavenTarget extends BaseTarget {
-          await git.push();
+  public async publish(_version: string, _revison: string): Promise<void> {
+    this.createUserGradlePropsFile();
+    if (isDryRun()) {
+      this.logger.info('[dry-run] Not uploading to Maven.');
+      return;
+    }
+    this.upload();
+    this.closeAndRelease();
+  }
+
+  /**
+   * Deploys to Maven Central the distribution packages.
+   * Note that after upload, this must be `closeAndRelease`.
+   */
+  public upload(): void {
+    const distributionsDirs = fs.readdirSync(
+      this.mavenConfig.distributionsPath
+    );
+    for (const distDir of distributionsDirs) {
+      const moduleName = path.parse(distDir).base;
+      const androidFile = this.getAndroidDistributionFile(distDir);
+      const targetFile = androidFile
+        ? androidFile
+        : path.join(distDir, `${moduleName}.jar`);
+      const javadocFile = path.join(distDir, `${moduleName}-javadoc.jar`);
+      const sourcesFile = path.join(distDir, `${moduleName}-sources.jar`);
+      const pomFile = path.join(distDir, 'pom-default.xml');
+
+      const command = this.getMavenUploadCmd(
+        targetFile,
+        javadocFile,
+        sourcesFile,
+        pomFile
+      );
+
+      withRetry(async () => {
+        exec(command, (error, _stdout, _stderr) => {
+          throw new Error(`Cannot upload ${distDir} to Maven:\n` + error);
+        });
+      });
+    }
+  }
+
+  /**
+   * Finishes the release flow.
+   */
+  public closeAndRelease(): void {
+    withRetry(async () => {
+      exec(
+        `./${this.mavenConfig.gradleCliPath} closeAndReleaseRepository`,
+        (error, _stdout, _stderr) => {
+          throw new Error(`Cannot close and release to Maven:\n` + error);
         }
-      },
-      true,
-      'craft-release-maven-'
+      );
+    });
+  }
+
+  /**
+   * Returns the path to the first Android distribution file, if any.
+   */
+  public getAndroidDistributionFile(
+    distributionDir: string
+  ): string | undefined {
+    const files = fs.readdirSync(distributionDir);
+    for (const filepath of files) {
+      const file = path.parse(filepath);
+      if (
+        file.ext === ANDROID_DIST_EXTENSION &&
+        file.base.includes(ANDROID_RELEASE_SUBSTR)
+      ) {
+        return filepath;
+      }
+    }
+    return undefined;
+  }
+
+  /**
+   * Returns the command to be executed, using the given parameters.
+   */
+  public getMavenUploadCmd(
+    targetFile: string,
+    javadocFile: string,
+    sourcesFile: string,
+    pomFile: string
+  ): string {
+    return (
+      `./${this.mavenConfig.mavenCliPath} gpg:sign-and-deploy-file ` +
+      `-Dfile=${targetFile} ` +
+      `-Dfiles=${javadocFile},${sourcesFile} ` +
+      `-Dclassifiers=javadoc,sources ` +
+      `-Dtypes=jar,jar ` +
+      `-DpomFile=${pomFile} ` +
+      `-DrepositoryId=${this.mavenConfig.mavenRepoId} ` +
+      `-Durl=${this.mavenConfig.mavenRepoUrl} ` +
+      `--settings ${this.mavenConfig.settingsPath} `
```

Comment:
This looks like a ticking bomb without proper shell escaping. The `exec` function should be doing this, why are you building a giant string instead of using a proper array so the function can do proper shell escaping?

Reply:
Still applies.